File: eidocell-backend/schemas/workspace/clusters.py
from typing import Literal

from pydantic import BaseModel, Field


class RunClusteringRequest(BaseModel):
    n_clusters: int = Field(10, ge=2, le=100)
    feature_method: str = "mobilenetv3"


class ClusteringScope(BaseModel):
    """Restrict clustering to a subset of samples.

    v1: a scoped run always replaces the session's existing clusters.
    """
    mode: Literal["all", "unlabeled", "class", "samples"] = "all"
    class_id: str | None = None
    sample_ids: list[str] | None = None


class RunClusteringPipelineRequest(BaseModel):
    """Full pipeline: feature extraction → dim reduction → clustering."""
    # Feature extraction
    feature_method: str = "mobilenetv3"

    # Dimensionality reduction (None = skip)
    dim_reduction_method: str | None = "pca"
    dim_reduction_params: dict = {}

    # Clustering
    clustering_method: str = "kmeans"
    n_clusters: int | None = Field(10, ge=2, le=100)
    clustering_params: dict = {}

    # Scope (defaults to all active samples)
    scope: ClusteringScope = Field(default_factory=ClusteringScope)


class SplitClusterRequest(BaseModel):
    n_sub_clusters: int = Field(2, ge=2, le=20)


class MergeClustersRequest(BaseModel):
    cluster_ids: list[str] = Field(..., min_length=2)


class AssignClustersToClassRequest(BaseModel):
    cluster_ids: list[str]
    class_id: str


class ClusterOut(BaseModel):
    id: str
    color: str
    sample_count: int
    labeled_count: int = 0
    quality_score: float | None = None
    feature_method: str | None = None

    model_config = {"from_attributes": True}


class EmbeddingPoint(BaseModel):
    sample_id: str
    x: float
    y: float
    cluster_id: str | None = None
    cluster_color: str | None = None


class ClusteringResult(BaseModel):
    clusters: list[ClusterOut]
    total_samples_clustered: int


class ClusteringPipelineResult(BaseModel):
    clusters: list[ClusterOut]
    total_samples_clustered: int
    embeddings: list[EmbeddingPoint]


class SplitClusterResult(BaseModel):
    new_clusters: list[ClusterOut]
    deleted_cluster_id: str


class MergeClustersResult(BaseModel):
    new_cluster: ClusterOut
    deleted_cluster_ids: list[str]


class ClusterAttributeStatistics(BaseModel):
    name: str
    mean: float | None = None
    std: float | None = None
    min: float | None = None
    max: float | None = None
    median: float | None = None


class ClusterStatistics(BaseModel):
    id: str
    color: str
    sample_count: int
    attributes: list[ClusterAttributeStatistics]

File: eidocell-ui/src/types/clusters.ts
export interface ClusterOut {
  id: string
  color: string
  sample_count: number
  labeled_count: number
  quality_score: number | null
  feature_method: string | null
}

export interface ClusteringResult {
  clusters: ClusterOut[]
  total_samples_clustered: number
}

export interface RunClusteringRequest {
  n_clusters: number
}

export type ClusteringScopeMode = 'all' | 'unlabeled' | 'class' | 'samples'

export interface ClusteringScope {
  mode: ClusteringScopeMode
  class_id?: string | null
  sample_ids?: string[] | null
}

export interface RunClusteringPipelineRequest {
  feature_method: string
  dim_reduction_method: string | null
  dim_reduction_params: Record<string, unknown>
  clustering_method: string
  n_clusters: number | null
  clustering_params: Record<string, unknown>
  scope?: ClusteringScope
}

export interface ClusterEmbeddingPoint {
  sample_id: string
  x: number
  y: number
  cluster_id: string | null
  cluster_color: string | null
}

export interface ClusteringPipelineResult {
  clusters: ClusterOut[]
  total_samples_clustered: number
  embeddings: ClusterEmbeddingPoint[]
}

export interface SplitClusterRequest {
  n_sub_clusters: number
}

export interface SplitClusterResult {
  new_clusters: ClusterOut[]
  deleted_cluster_id: string
}

export interface MergeClustersRequest {
  cluster_ids: string[]
}

export interface MergeClustersResult {
  new_cluster: ClusterOut
  deleted_cluster_ids: string[]
}

export interface AssignClustersToClassRequest {
  cluster_ids: string[]
  class_id: string
}

export type ClusterSortMode =
  | 'manual'
  | 'count_desc'
  | 'count_asc'
  | 'labeled_desc'
  | 'quality_desc'

export interface ClusterAttributeStatistics {
  name: string
  mean: number | null
  std: number | null
  min: number | null
  max: number | null
  median: number | null
}

export interface ClusterStatistics {
  id: string
  color: string
  sample_count: number
  attributes: ClusterAttributeStatistics[]
}
